File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/OSChannelTracker.java
package com.onesignal.influence;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.onesignal.OSLogger;
import com.onesignal.influence.model.OSInfluence;
import com.onesignal.influence.model.OSInfluenceChannel;
import com.onesignal.influence.model.OSInfluenceType;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

abstract public class OSChannelTracker {

    private static final String TIME = "time";

    protected OSLogger logger;
    @NonNull
    OSInfluenceDataRepository dataRepository;
    @Nullable
    OSInfluenceType influenceType;
    @Nullable
    JSONArray indirectIds;
    @Nullable
    String directId;

    OSChannelTracker(@NonNull OSInfluenceDataRepository dataRepository, OSLogger logger) {
        this.dataRepository = dataRepository;
        this.logger = logger;
    }

    public abstract String getIdTag();

    abstract OSInfluenceChannel getChannelType();

    abstract JSONArray getLastChannelObjectsReceivedByNewId(String id);

    abstract JSONArray getLastChannelObjects() throws JSONException;

    abstract int getChannelLimit();

    abstract int getIndirectAttributionWindow();

    abstract void saveChannelObjects(JSONArray channelObjects);

    abstract void initInfluencedTypeFromCache();

    abstract void addSessionData(@NonNull JSONObject jsonObject, OSInfluence influence);

    public abstract void cacheState();

    public void resetAndInitInfluence() {
        directId = null;
        indirectIds = getLastReceivedIds();
        influenceType = indirectIds.length() > 0 ? OSInfluenceType.INDIRECT : OSInfluenceType.UNATTRIBUTED;

        cacheState();
        logger.debug("OneSignal OSChannelTracker resetAndInitInfluence: " + getIdTag() + " finish with influenceType: " + influenceType);
    }

    /**
     * Get all received ids that may influence actions
     *
     * @return ids that happen between attribution window
     */
    public JSONArray getLastReceivedIds() {
        JSONArray ids = new JSONArray();
        try {
            JSONArray lastChannelObjectReceived = getLastChannelObjects();
            logger.debug("OneSignal ChannelTracker getLastReceivedIds lastChannelObjectReceived: " + lastChannelObjectReceived);

            long attributionWindow = getIndirectAttributionWindow() * 60 * 1_000L;
            long currentTime = System.currentTimeMillis();
            for (int i = 0; i < lastChannelObjectReceived.length(); i++) {
                JSONObject jsonObject = lastChannelObjectReceived.getJSONObject(i);
                long time = jsonObject.getLong(TIME);
                long difference = currentTime - time;

                if (difference <= attributionWindow) {
                    String id = jsonObject.getString(getIdTag());
                    ids.put(id);
                }
            }
        } catch (JSONException exception) {
            logger.error("Generating tracker getLastReceivedIds JSONObject ", exception);
        }

        return ids;
    }

    /**
     * Save state of last ids received
     */
    public void saveLastId(String id) {
        logger.debug("OneSignal OSChannelTracker for: " + getIdTag() + " saveLastId: " + id);
        if (id == null || id.isEmpty())
            return;

        JSONArray lastChannelObjectsReceived = getLastChannelObjectsReceivedByNewId(id);
        logger.debug("OneSignal OSChannelTracker for: " + getIdTag() + " saveLastId with lastChannelObjectsReceived: " + lastChannelObjectsReceived);

        try {
            JSONObject newInfluenceId = new JSONObject()
                    .put(getIdTag(), id)
                    .put(TIME, System.currentTimeMillis());
            lastChannelObjectsReceived.put(newInfluenceId);
        } catch (JSONException exception) {
            logger.error("Generating tracker newInfluenceId JSONObject ", exception);
            // We don't have new data, stop logic
            return;
        }

        int channelLimit = getChannelLimit();
        JSONArray channelObjectToSave = lastChannelObjectsReceived;

        // Only save the last ids without surpassing the limit
        // Always keep the max quantity of ids possible
        // If the attribution window increases, old ids might influence
        if (lastChannelObjectsReceived.length() > channelLimit) {
            int lengthDifference = lastChannelObjectsReceived.length() - channelLimit;
            // If min sdk is greater than KITKAT we can refactor this logic to removeObject from JSONArray
            channelObjectToSave = new JSONArray();
            for (int i = lengthDifference; i < lastChannelObjectsReceived.length(); i++) {
                try {
                    channelObjectToSave.put(lastChannelObjectsReceived.get(i));
                } catch (JSONException exception) {
                    logger.error("Before KITKAT API, Generating tracker lastChannelObjectsReceived get JSONObject ", exception);
                }
            }
        }

        logger.debug("OneSignal OSChannelTracker for: " + getIdTag() + " with channelObjectToSave: " + channelObjectToSave);
        saveChannelObjects(channelObjectToSave);
    }

    /**
     * Get the current session based on state + if outcomes features are enabled.
     */
    @NonNull
    public OSInfluence getCurrentSessionInfluence() {
        OSInfluence.Builder sessionInfluenceBuilder =
                OSInfluence.Builder.newInstance().setInfluenceType(OSInfluenceType.DISABLED);

        // Channel weren't init yet because application is starting
        if (influenceType == null)
            initInfluencedTypeFromCache();

        if (influenceType.isDirect()) {
            if (isDirectSessionEnabled()) {
                JSONArray directIds = new JSONArray().put(directId);
                sessionInfluenceBuilder = OSInfluence.Builder.newInstance()
                        .setIds(directIds)
                        .setInfluenceType(OSInfluenceType.DIRECT);
            }
        } else if (influenceType.isIndirect()) {
            if (isIndirectSessionEnabled()) {
                sessionInfluenceBuilder = OSInfluence.Builder.newInstance()
                        .setIds(indirectIds)
                        .setInfluenceType(OSInfluenceType.INDIRECT);
            }
        } else if (isUnattributedSessionEnabled()) {
            sessionInfluenceBuilder = OSInfluence.Builder.newInstance()
                    .setInfluenceType(OSInfluenceType.UNATTRIBUTED);
        }

        return sessionInfluenceBuilder
                .setInfluenceChannel(getChannelType())
                .build();
    }

    private boolean isDirectSessionEnabled() {
        return dataRepository.isDirectInfluenceEnabled();
    }

    private boolean isIndirectSessionEnabled() {
        return dataRepository.isIndirectInfluenceEnabled();
    }

    private boolean isUnattributedSessionEnabled() {
        return dataRepository.isUnattributedInfluenceEnabled();
    }

    @Nullable
    public OSInfluenceType getInfluenceType() {
        return influenceType;
    }

    public void setInfluenceType(@NonNull OSInfluenceType influenceType) {
        this.influenceType = influenceType;
    }

    @Nullable
    public JSONArray getIndirectIds() {
        return indirectIds;
    }

    public void setIndirectIds(@Nullable JSONArray indirectIds) {
        this.indirectIds = indirectIds;
    }

    @Nullable
    public String getDirectId() {
        return directId;
    }

    public void setDirectId(@Nullable String directId) {
        this.directId = directId;
    }

    @Override
    public String toString() {
        return "OSChannelTracker{" +
                "tag=" + getIdTag() +
                ", influenceType=" + influenceType +
                ", indirectIds=" + indirectIds +
                ", directId='" + directId + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        OSChannelTracker tracker = (OSChannelTracker) o;
        return influenceType == tracker.influenceType && tracker.getIdTag().equals(getIdTag());
    }

    @Override
    public int hashCode() {
        int result = influenceType.hashCode();
        result = 31 * result + getIdTag().hashCode();
        return result;
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/influence/data/OSChannelTracker.kt
package com.onesignal.influence.data

import com.onesignal.OSLogger
import com.onesignal.OSTime
import com.onesignal.influence.OSInfluenceConstants
import com.onesignal.influence.domain.OSInfluence
import com.onesignal.influence.domain.OSInfluenceChannel
import com.onesignal.influence.domain.OSInfluenceType
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

abstract class OSChannelTracker internal constructor(protected var dataRepository: OSInfluenceDataRepository, var logger: OSLogger, private var timeProvider: OSTime) {
    var influenceType: OSInfluenceType? = null
    var indirectIds: JSONArray? = null
    var directId: String? = null

    abstract val idTag: String
    abstract val channelType: OSInfluenceChannel

    @get:Throws(JSONException::class)
    abstract val lastChannelObjects: JSONArray
    abstract val channelLimit: Int
    abstract val indirectAttributionWindow: Int
    abstract fun getLastChannelObjectsReceivedByNewId(id: String?): JSONArray
    abstract fun saveChannelObjects(channelObjects: JSONArray)
    abstract fun initInfluencedTypeFromCache()
    abstract fun cacheState()
    abstract fun addSessionData(jsonObject: JSONObject, influence: OSInfluence)

    private val isDirectSessionEnabled: Boolean
        get() = dataRepository.isDirectInfluenceEnabled

    private val isIndirectSessionEnabled: Boolean
        get() = dataRepository.isIndirectInfluenceEnabled

    private val isUnattributedSessionEnabled: Boolean
        get() = dataRepository.isUnattributedInfluenceEnabled

    /**
     * Get the current session based on state + if outcomes features are enabled.
     */
    val currentSessionInfluence: OSInfluence
        get() {
            val sessionInfluence = OSInfluence(channelType, OSInfluenceType.DISABLED, null)
            // Channel weren't init yet because application is starting
            if (influenceType == null) initInfluencedTypeFromCache()

            val currentInfluenceType = influenceType ?: OSInfluenceType.DISABLED

            if (currentInfluenceType.isDirect()) {
                if (isDirectSessionEnabled) {
                    sessionInfluence.apply {
                        ids = JSONArray().put(this@OSChannelTracker.directId)
                        influenceType = OSInfluenceType.DIRECT
                    }
                }
            } else if (currentInfluenceType.isIndirect()) {
                if (isIndirectSessionEnabled) {
                    sessionInfluence.apply {
                        ids = this@OSChannelTracker.indirectIds
                        influenceType = OSInfluenceType.INDIRECT
                    }
                }
            } else if (isUnattributedSessionEnabled) {
                sessionInfluence.apply {
                    influenceType = OSInfluenceType.UNATTRIBUTED
                }
            }

            return sessionInfluence
        }

    /**
     * Get all received ids that may influence actions
     * @return ids that happen between attribution window
     */
    val lastReceivedIds: JSONArray
        get() {
            val ids = JSONArray()
            try {
                val lastChannelObjectReceived = lastChannelObjects
                logger.debug("OneSignal ChannelTracker getLastReceivedIds lastChannelObjectReceived: $lastChannelObjectReceived")
                val attributionWindow = indirectAttributionWindow * 60 * 1000L
                val currentTime = timeProvider.currentTimeMillis
                for (i in 0 until lastChannelObjectReceived.length()) {
                    val jsonObject = lastChannelObjectReceived.getJSONObject(i)
                    val time = jsonObject.getLong(OSInfluenceConstants.TIME)
                    val difference = currentTime - time
                    if (difference <= attributionWindow) {
                        val id = jsonObject.getString(idTag)
                        ids.put(id)
                    }
                }
            } catch (exception: JSONException) {
                logger.error("Generating tracker getLastReceivedIds JSONObject ", exception)
            }
            return ids
        }

    fun resetAndInitInfluence() {
        directId = null
        indirectIds = lastReceivedIds
        influenceType = if (indirectIds?.length() ?: 0 > 0) OSInfluenceType.INDIRECT else OSInfluenceType.UNATTRIBUTED
        cacheState()
        logger.debug("OneSignal OSChannelTracker resetAndInitInfluence: $idTag finish with influenceType: $influenceType")
    }

    /**
     * Save state of last ids received
     */
    fun saveLastId(id: String?) {
        logger.debug("OneSignal OSChannelTracker for: $idTag saveLastId: $id")
        if (id == null || id.isEmpty()) return

        val lastChannelObjectsReceived = getLastChannelObjectsReceivedByNewId(id)
        logger.debug("OneSignal OSChannelTracker for: $idTag saveLastId with lastChannelObjectsReceived: $lastChannelObjectsReceived")

        try {
            timeProvider.run {
                JSONObject()
                        .put(idTag, id)
                        .put(OSInfluenceConstants.TIME, currentTimeMillis)
            }.also { newInfluenceId ->
                lastChannelObjectsReceived.put(newInfluenceId)
            }
        } catch (exception: JSONException) {
            logger.error("Generating tracker newInfluenceId JSONObject ", exception)
            // We don't have new data, stop logic
            return
        }

        var channelObjectToSave = lastChannelObjectsReceived
        // Only save the last ids without surpassing the limit
        // Always keep the max quantity of ids possible
        // If the attribution window increases, old ids might influence
        if (lastChannelObjectsReceived.length() > channelLimit) {
            val lengthDifference = lastChannelObjectsReceived.length() - channelLimit
            // If min sdk is greater than KITKAT we can refactor this logic to removeObject from JSONArray
            channelObjectToSave = JSONArray()
            for (i in lengthDifference until lastChannelObjectsReceived.length()) {
                try {
                    channelObjectToSave.put(lastChannelObjectsReceived[i])
                } catch (exception: JSONException) {
                    logger.error("Generating tracker lastChannelObjectsReceived get JSONObject ", exception)
                }
            }
        }
        logger.debug("OneSignal OSChannelTracker for: $idTag with channelObjectToSave: $channelObjectToSave")
        saveChannelObjects(channelObjectToSave)
    }

    override fun toString(): String {
        return "OSChannelTracker{" +
                "tag=" + idTag +
                ", influenceType=" + influenceType +
                ", indirectIds=" + indirectIds +
                ", directId=" + directId +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val tracker = other as OSChannelTracker
        return influenceType === tracker.influenceType && tracker.idTag == idTag
    }

    override fun hashCode(): Int {
        var result = influenceType.hashCode()
        result = 31 * result + idTag.hashCode()
        return result
    }
}